components/ws-manager-mk2/controllers/workspace_controller.go

```diff
@@ -221,6 +219,55 @@ func (r *WorkspaceReconciler) actOnStatus(ctx context.Context, workspace *worksp
 				r.Recorder.Event(workspace, corev1.EventTypeNormal, "Creating", "")
 			}
 
+		case workspace.Status.Phase == workspacev1.WorkspacePhaseStopped && workspace.IsConditionTrue(workspacev1.WorkspaceConditionPodRejected):
+			if workspace.Status.PodRecreated > r.Config.PodRecreationMaxRetries {
+				workspace.Status.SetCondition(workspacev1.NewWorkspaceConditionPodRejected(fmt.Sprintf("Pod reached maximum recreations %d, failing", workspace.Status.PodRecreated), metav1.ConditionFalse))
+				return ctrl.Result{Requeue: true}, nil // requeue so we end up in the "Stopped" case below
+			}
+			log.WithValues("PodStarts", workspace.Status.PodStarts, "PodRecreated", workspace.Status.PodRecreated, "Phase", workspace.Status.Phase).Info("trigger pod recreation")
+
+			// Make sure to wait for "recreationTimeout" before creating the pod again
+			if workspace.Status.PodDeletionTime == nil {
+				log.Info("want to wait for pod recreation timeout, but podDeletionTime not set (yet)")
+				return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+			}
+
+			recreationTimeout := r.podRecreationTimeout()
+			podDeletionTime := workspace.Status.PodDeletionTime.Time
+			waitTime := time.Until(podDeletionTime.Add(recreationTimeout))
+			if waitTime > 0 {
+				log.WithValues("waitTime", waitTime).Info("waiting for pod recreation timeout")
+				return ctrl.Result{Requeue: true, RequeueAfter: waitTime}, nil
+			}
+			log.WithValues("waitedTime", waitTime.Abs().String()).Info("waited for pod recreation timeout")
+
+			// Must persist the modification pod starts, and ensure we retry on conflict.
+			// If we fail to persist this value, it's possible that the Pod gets recreated endlessly
+			// when the workspace stops, due to PodStarts still being 0 when the original Pod
+			// disappears.
+			// Use a Patch instead of an Update, to prevent conflicts.
+			patch := client.MergeFrom(workspace.DeepCopy())
+
+			// Reset status
+			sc := workspace.Status.DeepCopy()
+			workspace.Status = workspacev1.WorkspaceStatus{}
+			workspace.Status.Phase = workspacev1.WorkspacePhasePending
+			workspace.Status.OwnerToken = sc.OwnerToken
+			workspace.Status.PodStarts = sc.PodStarts
+			workspace.Status.PodRecreated = sc.PodRecreated + 1
+			workspace.Status.SetCondition(workspacev1.NewWorkspaceConditionPodRejected(fmt.Sprintf("Recreating pod... (%d retry)", workspace.Status.PodRecreated), metav1.ConditionFalse))
+
+			if err := r.Status().Patch(ctx, workspace, patch); err != nil {
```

Comment:
I'm not super convinced this should be a `Patch`, an `Update` would catch acting on an out-of-date status by returning a conflict error

Reply:
Hm, with how the current code works, I think I agree. Will try it out. :+1:

Reply:
`Update` worked flawlessly, will keep it!